src/MongoDB.Driver.Core/Core/Operations/AsyncCursor.cs

```diff
@@ -273,39 +273,55 @@ private BsonDocument CreateKillCursorsCommand()
         private CursorBatch<TDocument> ExecuteGetMoreCommand(IChannelHandle channel, CancellationToken cancellationToken)
         {
             var command = CreateGetMoreCommand();
-            var result = channel.Command<BsonDocument>(
-                _channelSource.Session,
-                null, // readPreference
-                _collectionNamespace.DatabaseNamespace,
-                command,
-                null, // commandPayloads
-                NoOpElementNameValidator.Instance,
-                null, // additionalOptions
-                null, // postWriteAction
-                CommandResponseHandling.Return,
-                __getMoreCommandResultSerializer,
-                _messageEncoderSettings,
-                cancellationToken);
+            BsonDocument result;
+            try
+            {
+                result = channel.Command<BsonDocument>(
+                    _channelSource.Session,
+                    null, // readPreference
+                    _collectionNamespace.DatabaseNamespace,
+                    command,
+                    null, // commandPayloads
+                    NoOpElementNameValidator.Instance,
+                    null, // additionalOptions
+                    null, // postWriteAction
+                    CommandResponseHandling.Return,
+                    __getMoreCommandResultSerializer,
+                    _messageEncoderSettings,
+                    cancellationToken);
+            }
+            catch (MongoCommandException ex) when (IsMongoCursorNotFoundException(ex))
+            {
+                throw new MongoCursorNotFoundException(channel.ConnectionDescription.ConnectionId, _cursorId, command);
+            }
```

Comment:
This seems like the wrong level to be catching this (too high).

This should be caught somewhere much deeper and converted there.

Reply:
Maybe in `ExceptionMapper.Map`?

Reply:
> Maybe in ExceptionMapper.Map?

I have considered this option, but this exception refers to "cursor" area and additionally `_cursorId` is only available on `ExecuteGetMoreCommand()`, which makes it ineffective to capture at lower level.

Not done.

Reply:
Does the error message from the server not mention the cursor id?

It would be better if exception mapping were kept centralized in `ExceptionMapper`.

Reply:
I think the server response contains the cursorId and I also agree that it will be easier to have all this logic in one place, but I think it's not really correct that we need to parse the operation details related to higher levels on the lowest wireprotocol level. After some time, it can lead to the situation when this method will know all details about higher levels and as a result, will be too complicated.
So I think it will be not the best design

Reply:
Thanks for the discussion. I agree now that this is the right place for this check.